lib/utils/logDeprecation.js

```diff
@@ -0,0 +1,44 @@
+'use strict';
+const chalk = require('chalk');
+const memoizee = require('memoizee');
+
+const loggedDeprecations = {};
+
+function consoleLog(...args) {
+  console.log(...args); // eslint-disable-line no-console
+}
+
+function extractCodes(codesStr) {
+  if (!codesStr) {
+    return [];
+  }
+  return codesStr.split(',').reduce((acc, code) => {
+    acc[code] = true;
+    return acc;
+  }, {});
+}
+
+const prepareCodes = memoizee((serviceConfigCodesStr, envCodesStr) => {
+  const serviceConfigCodes = extractCodes(serviceConfigCodesStr);
+  const envCodes = extractCodes(envCodesStr);
+  const disabledCodes = Object.assign({}, envCodes, serviceConfigCodes);
+  return disabledCodes;
+});
+
+function writeDeprecation(message, docURL) {
+  consoleLog(`${chalk.red(message)}\n${chalk.dim(`More Info: ${docURL}`)}`);
+}
+
+function logDeprecation(code, message, docURL, serviceConfig) {
+  const disabledCodes = prepareCodes(
+    serviceConfig && serviceConfig.disabledCodes,
```

Comment:
I would create one dedicated function to resolve deprecated codes from `serviceConfig`. I wouldn't mix it with resolution from `process.env` as it's different, and settings from `process.env` can be resolved upfront in module body

Reply:
The idea was to keep the codes preparation separate and explicit so you know what was prepared for this call explicitly. The env codes and serviceCodes (if they exist) were extracted in single call to make it easier to trace what was extracted
I assume `serviceConfig.disabledCodes` will always return either an empty Set or a Set with codes to   disable. So I can use it without taking too much care about it being sent or not.

I separated them, but why do you want to keep the serviceConfig codes resolution and checking that separate?

Reply:
> The env codes and serviceCodes (if they exist) were extracted in single call to make it easier to trace what was extracted

I understand the intention, still both values are provided at different points in time, and format of both differs. So resolving both with one function, doesn't seem clean to me.

I believe that `process.env` can be resolved upfront in module body, so we have upfront (for function processing) ready _set_ of codes deprecated by env settings. I think it'll make further processing of that cleaner.

`process.env` format is in all cases it's a string, which eventually may contain  multiple coma separated values. While in case of `serviceConfig`, it can be either string (singular value) or array. So resolution of both should in any case be done with different logic (different function)

Reply:
>  While in case of serviceConfig, it can be either string (singular value) or array

So it will be provided like this?
```yml
provider: aws
disabledCodes:
  - code1
  - code2
```

I thought it will be comma separated values like the env var
```yml
provider: aws
disabledCodes: code1,code2,code3
```

Or it will be processed by service initialisation? Because I assumed it should match how it's provided to the Env var

Reply:
In https://github.com/serverless/serverless/issues/7422 I've proposed to name it `disabledDeprecations`, and concerning format we should respect convention in  how single and multiple valuas are configured in YAML (which translates to JSON).

Still I think it's fine to support just one format (an array), at least singular string will look confusing by `disabledDeprecations` which suggests plural.

So let's support just:

```yaml
disabledDeprecations:
  - CODE1
  - CODE2
```

Reply:
And for `disabledDeprecations` let's make a one test that involves `runServelress` (only that way we should test input coming via `serviceConfig`

Reply:
> In #7422 I've proposed to name it disabledDeprecations, and concerning format we should respect convention in how single and multiple valuas are configured in YAML (which translates to JSON).

Sorry for that, I was trying to get sth submitted to discuss before finalizing the implementation

> Still I think it's fine to support just one format (an array), at least singular string will look confusing by disabledDeprecations which suggests plural.

You always get the request to support `'*'` as a single value instead of a single array item. It doesn't matter here?

Reply:
> You always get the request to support '*' as a single value instead of a single array item. It doesn't matter here?

Good point! I forgot about that. In that case let's support also single value notation where a _string_ is assigned to `disabledDeprecations` directly.

Also I wonder, maybe more correct name would be `disabledDeprecationNotifications` (at least `disabledDeprecations` suggest we _disable_ deprecations which is not perfectly correct). What do you think?

Reply:
I added the support for single value either way until you reply